File: tsp/grafo.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
	int valor;
	int peso;
} Elemento;

typedef struct {
	int id;
	Elemento lista_adj[100];
	int tamanho_lista;
} Vertice;

Elemento novo_elemento(int valor, int peso){
	Elemento novo;
	novo.valor = valor;
	novo.peso = peso;
	return novo;
}

int fat(int num)
{
	if(num <= 1){
		return 1;
	}else{
		return num * fat(num-1);
	}
}

void addAresta(Vertice * v, int x, int y, int peso){
	Elemento novo = novo_elemento(y, peso);

	v[x].lista_adj[ v[x].tamanho_lista++ ] = novo;

	novo = novo_elemento(x, peso);
	v[y].lista_adj[ v[y].tamanho_lista++ ] = novo;
}

//int * calcula_rota(Vertice * v, int n, int root);
void calcula_rota(Vertice * v, int n, int root);
void calcula_distancia(Vertice * v, int * visitados, int elemento, int dist_atual, int n, int root);

int menor_distacia = 9999;
int * melhor_rota;

int main()
{
	int n, x, y, peso, i, lim;

	// scanf("%d", &n);
	n = 4;

	Vertice * vertices = malloc(sizeof(Vertice)*(n));

	lim = fat(n-1);

	addAresta(vertices, 0, 1, 3);
	addAresta(vertices, 0, 2, 4);
	addAresta(vertices, 0, 3, 5);
	addAresta(vertices, 1, 2, 5);
	addAresta(vertices, 1, 3, 4);
	addAresta(vertices, 2, 3, 3);

	/* for(i = 0; i < lim; i++){
		scanf("%d %d %d", &x, &y, &peso);

		addAresta(vertices, x, y, peso);
		addAresta(vertices, y, x, peso);
	} */

	int * rota = malloc(sizeof(int)*(n+1));

	calcula_rota(vertices, n, 0);

	printf("\nMenor distancia: %d", menor_distacia);

	printf("\n");
	return 0;
}

//int * calcula_rota(Vertice * v, int n, int root)
void calcula_rota(Vertice * v, int n, int root)
{
	//int * rota = malloc(sizeof(int)*(n+1));
	int * visitados = malloc(sizeof(int)*n);
	int elemento, i, dist, dist_atual;
	melhor_rota = malloc(sizeof(int)*(n+1));

	visitados[root] = 1;

	for(i = 0; i < n; i++){
		elemento = v[root].lista_adj[i].valor;
		dist = v[root].lista_adj[i].peso;
		
		dist_atual = 0;

		calcula_distancia(v, visitados, elemento, dist_atual+dist, n, root);
	}

	//return rota;
}

void calcula_distancia(Vertice * v, int * visitados, int elemento, int dist_atual, int n, int root){

	int i, this_el, dist;

	visitados[elemento] = 1;

	int flag = 1;

	int tam = v[elemento].tamanho_lista;

	for(i = 0; i < tam; i++){
		this_el = v[elemento].lista_adj[i].valor;

		if(visitados[ this_el ] == 0){
			flag = 0;
			dist = v[elemento].lista_adj[i].peso;
			calcula_distancia(v, visitados, this_el, dist_atual+dist, n, root);
		}
	}

	if(flag == 1){

		dist = v[root].lista_adj[elemento].peso + dist_atual;

		if(dist < menor_distacia){
			menor_distacia = dist;
		}
	}
}

File: tsp/node/grafo.js
const Vertice = require("./vertice.js");

class Grafo {

	constructor(num_vertices){
		this.vertices = [];
		this.num_vertices = num_vertices;
		this.rota_otima = [];
		this.menor_distancia = 9999;

		for(let i = 0; i < num_vertices; i++){
			this.vertices.push(new Vertice());
		}
	}

	addAresta(x, y, peso){
		this.vertices[x].addAresta(y, peso);

		this.vertices[y].addAresta(x, peso);
	}

	mostraGrafo() {
        for (let i = 0; i < this.num_vertices; i++) {
            console.log(`${i} -> ${this.vertices[i].getListaAdj()}`);
        }
    }

	getDistance(x, y){
		let lista = this.vertices[x].getCompleteList();
		let dist = 0;

		lista.forEach((e)=>{
			if(e.valor == y){
				dist = e.peso;
			}
		});
		return dist;
	}

	getDistancia(x, y){
		let tam = this.vertices[x].getTamanhoLista();
		for(let i = 0; i < tam; i++){
			let elemento = this.vertices[x].getListaAdj()[i];

			if(elemento == y){
				return this.vertices[x].getListaPeso()[i];
			}
		}

	}

	calcula_distancia(route){
		let distancia = 0;
		let tam = route.length;

		for(let i = 0, j = 1; i < tam-1 && j < tam; i++, j++){
			distancia += this.getDistance(route[i], route[j]);
		}
		return distancia;
	}

	tsp(origin){
		var indice = 0;
		var route = [];
		var visitados = [];
		for(let i = 0; i < this.num_vertices; i++){
			visitados.push(0);
			route.push(0);
		}

		route.push(origin)

		route[indice] = origin;
		visitados[origin] = 1;

		for(let i = 0; i < this.vertices[origin].getTamanhoLista(); i++){
			let elemento = this.vertices[origin].getListaAdj()[i]
			this.calcula(elemento, visitados, route, indice+1);
		}

		return {
			'rota':this.rota_otima,
			'distancia':this.menor_distancia
		}
	}

	calcula(elemento, visitados, route, indice){

		route[indice] = elemento;
		visitados[elemento] = 1;
		
		if(indice == this.num_vertices-1){
			let dist = this.calcula_distancia(route);
			if(dist < this.menor_distancia){
				this.menor_distancia = dist;
				for(let i = 0; i <= this.num_vertices; i++){
					this.rota_otima[i] = route[i]
				}
			}
		}else{

			for(let i = 0; i < this.vertices[elemento].getTamanhoLista(); i++){
				let el = this.vertices[elemento].getListaAdj()[i];
				if(visitados[el] == 0){
					this.calcula(el, visitados, route, indice+1);
				}
			}
		}

		visitados[elemento] = 0
	}

	caixeiro(origin){
		this.rota_otima = [...Array(this.num_vertices+1).fill(0)];
		this.menor_distancia = 999999;

		var indice = 0;
		const route = [...Array(this.num_vertices+1).fill(0)];
		const visitados = [...Array(this.num_vertices).fill(0)];
		
		visitados[origin] = 1;
		route[0] = origin;
		route[this.num_vertices] = origin;

		var tam = this.vertices[origin].getTamanhoLista();

		for(let i = 0; i < tam; i++){
			let elemento = this.vertices[origin].getListaAdj()[i]
			let distancia = this.vertices[origin].getListaPeso()[i]

			this.viajante(origin, visitados, route, indice+1, elemento, distancia);
		}

		return {
			'rota':this.rota_otima,
			'peso':this.menor_distancia
		}
	}

	viajante(origin, visitados, route, indice, elemento, distancia){

		if(distancia > this.menor_distancia) return;

		visitados[elemento] = 1;
		route[indice] = elemento;

		let tam = this.vertices[elemento].getTamanhoLista();

		for(let i = 0; i < tam; i++){
			let el_lista = this.vertices[elemento].getListaAdj()[i];
			let el_dist = this.vertices[elemento].getListaPeso()[i];

			if(visitados[el_lista] == 0){
				this.viajante(origin, visitados, route, indice+1, el_lista, distancia+el_dist);
			}
		}

		if(indice == this.num_vertices-1){
			let dist = distancia + this.getDistancia(origin, elemento);

			if(dist <= this.menor_distancia){
				this.menor_distancia = dist;
				for(let i = 0; i <= this.num_vertices; i++){
					this.rota_otima[i] = route[i]
				}
			}
		}

		visitados[elemento] = 0;
	}
}

module.exports = Grafo;
